src/twisted/test/test_defer.py

```diff
@@ -645,6 +645,150 @@ def test_cancelGatherResultsWithAllDeferredsCallback(self):
+        """
+        L{defer.extractCurrentResult} raises a L{defer.NoCurrentResult}
+        exception which is not caught when catching L{Exception}.
+        """
+        def inner(d):
+            try:
+                defer.extractCurrentResult(d)
+            except Exception:
+                return "error"
+            return "success"
+
+        d = defer.Deferred()
+        self.assertRaises(defer.NoCurrentResult, inner, d)
+
+        # In order to get 100% patch coverage we need to excerise the pass
+        # statement. This is a bit silly because we're basically testing our
+        # tests and the fact that this code does not run is proof that the unit
+        # under test is correctly functioning. However, we'll jump around on
+        # one foot to satisify the policy.
+        d = defer.Deferred()
+        d.errback(Exception())
+        self.assertEqual(inner(d), "error")
+        d = defer.Deferred()
+        d.callback(None)
+        self.assertEqual(inner(d), "success")
+
+
+    def test_extractCurrentResultPreservesCallbackChain(self):
+        """
+        L{defer.extractCurrentResult} chains the value that will be passed into
+        the next function of the callback chain if passthrough=True.
+        """
+        result1 = object()
+        result2 = object()
+
+        d = defer.Deferred()
+        d.callback(result1)
+
+        self.assertIs(defer.extractCurrentResult(d, passthrough=True), result1)
+
+        def cb(r):
+            self.assertIs(r, result1)
+            return result2
+
+        d.addCallback(cb)
+
+        self.assertIs(defer.extractCurrentResult(d), result2)
+
+
+    def test_extractCurrentResultRaisesErrorIfPending(self):
+        """
+        L{defer.extractCurrentResult} raises an error if there currently any
+        pending work to be done on the L{Deferred}.
+        """
+        d = defer.Deferred()
+
+        self.assertRaises(defer.NoCurrentResult, defer.extractCurrentResult, d)
+
+
+    def test_extractCurrentResultRaisesErrorIfPendingChainedDeferred(self):
+        """
+        L{defer.extractCurrentResult} raises an error if there is currently any
+        pending work to be done on the L{Deferred} if it is via a chained
+        L{Deferred}.
+        """
+        result = object()
+
+        d = defer.Deferred()
+        d.callback(result)
+
+        self.assertIs(defer.extractCurrentResult(d), result)
+
+        d2 = defer.Deferred()
+        d.addCallback(lambda _: d2)
+
+        self.assertRaises(defer.NoCurrentResult, defer.extractCurrentResult, d)
+
+
+    def test_extractCurrentResultDoesntPassthroughResults(self):
+        """
+        L{defer.extractCurrentResult} does not pass through results.
+        """
+        result = object()
+
+        d = defer.Deferred()
+        d.callback(result)
+
+        self.assertIs(defer.extractCurrentResult(d), result)
+        self.assertIsNot(defer.extractCurrentResult(d), result)
```

Comment:
Assert that the result becomes `None` perhaps?  Asserting it's not the result leaves the field wide open for what the behavior actually is.  I don't see any reason to let the test suite leave the intent this wide.